substrate/primitives/core/src/address_uri.rs

```diff
@@ -0,0 +1,258 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Little util for parsing an address URI. Replaces regular expressions.
+
+#[derive(Debug, PartialEq)]
+/// A container for results of parsing the address uri string.
+pub struct AddressUri<'a> {
+	pub ss58: Option<&'a str>,
+	pub paths: Vec<&'a str>,
+	pub pass: Option<&'a str>,
+}
+
+impl<'a> AddressUri<'a> {
+	fn extract_prefix(input: &mut &'a str, is_allowed: &dyn Fn(char) -> bool) -> Option<&'a str> {
+		let output = input.trim_start_matches(is_allowed);
+		let prefix_len = input.len() - output.len();
+		let prefix = if prefix_len > 0 { Some(&input[..prefix_len]) } else { None };
+		*input = output;
+		prefix
+	}
+
+	/// Parses the given string.
+	///
+	/// Intended to be equivalent of:
+	/// Regex::new(r"^(?P<phrase>[a-zA-Z0-9 ]+)?(?P<path>(//?[^/]+)*)(///(?P<password>.*))?$")
+	pub fn parse(mut input: &'a str) -> Option<Self> {
+		let ss58 = Self::extract_prefix(&mut input, &|ch: char| {
+			ch.is_ascii_digit() || ch.is_ascii_alphabetic() || ch == ' '
+		});
+
+		let mut pass = None;
+		let mut paths = Vec::new();
+		while !input.is_empty() {
+			input = if let Some(mut maybe_pass) = input.strip_prefix("///") {
+				pass = match Self::extract_prefix(&mut maybe_pass, &|ch: char| ch != '\n') {
+					Some(pass) => Some(pass),
+					None => Some(""),
+				};
```

Comment:
```suggestion
				pass = Self::extract_prefix(&mut maybe_pass, &|ch: char| ch != '\n');
```

`pass` is an `Option`?

Reply:
Yes. But regex will return `Some("")` if intput ends with  `"///"` (e.g. `hello world///`), and there is no password.
I added this case to all seamless testing agains regex.